src/USER-HSMA/Install.sh

```diff
@@ -0,0 +1,56 @@
+# Install/unInstall package files in LAMMPS
+# mode = 0/1/2 for uninstall/install/update
+
+mode=$1
+
+# enforce using portable C locale
+LC_ALL=C
+export LC_ALL
+
+# arg1 = file, arg2 = file it depends on
+
+action () {
+  if (test $mode = 0) then
+    rm -f ../$1
+  elif (! cmp -s $1 ../$1) then
+    if (test -z "$2" || test -e ../$2) then
+      cp $1 ..
+      if (test $mode = 2) then
+        echo "  updating src/$1"
+      fi
+    fi
+  elif (test -n "$2") then
+    if (test ! -e ../$2) then
+      rm -f ../$1
+    fi
+  fi
+}
+
+# all package files with no dependencies
+
+for file in *.cpp *.h; do
+  test -f ${file} && action $file
+done
+
+# edit 2 Makefile.package files to include/exclude package info
+
+if (test $1 = 1) then
+
+  if (test -e ../Makefile.package) then
+    sed -i -e 's/`.*hsma.*` //' ../Makefile.package
+    sed -i -e 's/[^ \t]*hsma[^ \t]* //' ../Makefile.package
+        sed -i -e '2a SOURCE=$(wildcard ../USER-HSMA/OFile/* .o) ' ../Makefile.package
```

Comment:
Please do not include binary files in your pull request. If you need additional files with support functions, you can provide them as a separate library (to be compiled from their sources) in the lib folder. LAMMPS is compiled and used on multiple CPU architectures (x86 32-bit, x86 64-bit, ARM, IBM Power, etc.) and Operating systems (Linux, MacOS, FreeBSD, Windows) so all contributions should be as portable as possible.

Reply:
Thank you for your good suggestions. In our codes, [the FMM package](https://github.com/flatironinstitute/FMM3D) which is also released as open source developed by Greengard's group is used. I am sorry that I don't know how to compile them as a separate library. Can you provide some reference?

Reply:
Please see the existing folders in the "lib" directories. For building with the traditional make you need to make a copy and adapt one of the existing Install.py files and then also provide one or more Makefile.lammps file with the settings needed to add to the main make for linking to the library (depending on the needs of individual compilers). The Install.py file may be called from the main makefile to download and build the external package files. Mind you, this also needs your package to be added to the proper variables in src/Makefile (to indicate that your package is a user package and requires a library) and you need to make changes to src/USER-HSMA/Install.sh to import settings from lib/hsma/Makefile.lammps.

For CMake you need to create/adapt a package specific file in cmake/Modules/Packages that also allows downloading and building the support library and enables the necessary settings and then add/configure the package in cmake/CMakeLists.txt. You can look at the existing packages for examples of how to do that.

Please also remember that you need to update the manual in doc/src accordingly with the suitable instructions and explanations. I've pointed out the relevant files in a different comment.

Please note that currently you package is not included into the automatic build tests, but it will be added and has to pass building on our test platforms with both traditional makefiles and cmake.

Reply:
Thank you for these remarks which are very helpful. We will try to fix these problems in a few days.